src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,440 +233,82 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        protected void IncreaseBatchSize()
         {
-            switch (this.batchReaderState)
+            this.currentBatchSize++;
+
+            if (this.currentBatchSize > this.inputContext.MessageReaderSettings.MessageQuotas.MaxPartsPerBatch)
             {
-                case ODataBatchReaderState.Initial:
-                    // If we find an end boundary when in state 'Initial' it means that we
-                    // have an empty batch. The next state will be 'Completed'.
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Operation:
-                    // If we find an end boundary in state 'Operation' we have finished
-                    // processing an operation and found the end boundary of either the
-                    // current changeset or the batch.
-                    return this.batchStream.ChangeSetBoundary == null
-                        ? ODataBatchReaderState.Completed
-                        : ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetStart:
-                    // If we find an end boundary when in state 'ChangeSetStart' it means that
-                    // we have an empty changeset. The next state will be 'ChangeSetEnd'
-                    return ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetEnd:
-                    // If we are at the end of a changeset and find an end boundary
-                    // we reached the end of the batch
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Completed:
-                    // We should never get here when in Completed state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Completed));
-
-                case ODataBatchReaderState.Exception:
-                    // We should never get here when in Exception state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Exception));
-
-                default:
-                    // Invalid enum value
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_UnknownValue));
+                throw new ODataException(Strings.ODataBatchReader_MaxBatchSizeExceeded(this.inputContext.MessageReaderSettings.MessageQuotas.MaxPartsPerBatch));
             }
         }
 
         /// <summary>
-        /// Reads the next part from the batch message payload.
+        /// Increases the size of the current change set; throws if the allowed limit is exceeded.
         /// </summary>
-        /// <returns>true if more information was read; otherwise false.</returns>
-        private bool ReadSynchronously()
+        protected void IncreaseChangesetSize()

```

Comment:
>IncreaseChangesetSize [](start = 23, length = 21)

Can IncreaseBatchSize and IncreaseChangesetSize be made private, and called in CreateOperationRequestMessage in the base class?

Reply:
They both are invoked only from ReadImplementation().

---
In reply to: [141147205](https://github.com/OData/odata.net/pull/960#discussion_r141147205) [](ancestors = 141147205)